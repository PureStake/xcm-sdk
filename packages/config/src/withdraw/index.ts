export * from './withdraw';
export * from './withdraw.interfaces';
