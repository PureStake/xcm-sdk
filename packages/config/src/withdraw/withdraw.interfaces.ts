import { BalanceConfig } from '../balance';
import { Asset } from '../constants';
import { ChainConfig } from '../interfaces';

export type WithdrawConfig<Assets extends Asset> =
  WithdrawXTokensConfig<Assets>;

export interface WithdrawXTokensConfig<Assets extends Asset> {
  balance: BalanceConfig<Assets>;
  destination: ChainConfig;
  existentialDeposit: number | bigint;
  feePerWeight: number;
  weight: number | bigint;
  getParams: (account: string) => DestinationML;
}

export type DestinationML = [
  /**
   * 1 - if transaction is going through a relay chain
   */
  1,
  (
    | [
        /**
         * example '0x00000007DC'
         * 7DC - parachain id in hex
         * can be found here:
         *   - https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fkusama-rpc.polkadot.io#/parachains
         *   - https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc.polkadot.io#/parachains
         */
        string,
        /**
         * example '0x01%account%00',
         * enum = 01 (AccountId32)
         * networkId = 00 (any)
         */
        string,
      ]
    | [
        /**
         * example '0x01%account%00',
         * enum = 01 (AccountId32)
         * networkId = 00 (any)
         */
        string,
      ]
  ),
];

export interface WithdrawXTokensOptions<Assets extends Asset> {
  balance: BalanceConfig<Assets>;
  destination: ChainConfig;
  existentialDeposit?: number | bigint;
  feePerWeight: number;
  weight?: number | bigint;
}
