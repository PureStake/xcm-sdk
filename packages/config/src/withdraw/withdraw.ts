import { u8aToHex } from '@polkadot/util';
import { decodeAddress } from '@polkadot/util-crypto';
import { AssetSymbol } from '../constants';
import {
  WithdrawXTokensConfig,
  WithdrawXTokensOptions,
} from './withdraw.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function createWithdrawBuilder<
  Symbols extends AssetSymbol = AssetSymbol,
>() {
  return {
    xTokens: (options: WithdrawXTokensOptions<Symbols>) =>
      xTokens<Symbols>(options),
  };
}

function xTokens<Symbols extends AssetSymbol = AssetSymbol>({
  balance,
  destination,
  feePerWeight,
  sourceMinBalance,
  xcmFeeAsset,
  weight = 4_000_000_000,
}: WithdrawXTokensOptions<Symbols>): WithdrawXTokensConfig<Symbols> {
  return {
    balance,
    destination,
    feePerWeight,
    sourceMinBalance,
    xcmFeeAsset,
    weight,
    getParams: (account: string, usesEthereumAccounts = false) => {
      const { parachainId } = destination;
      // 01: AccountId32
      // 03: AccountKey20
      // https://docs.moonbeam.network/builders/interoperability/xcm/xc20/xtokens/#building-the-precompile-multilocation
      const accountType = usesEthereumAccounts ? '03' : '01';

      const acc = `0x${accountType}${u8aToHex(
        decodeAddress(account),
        -1,
        false,
      )}00`;

      return [
        1,
        parachainId ? [`0x0000000${parachainId.toString(16)}`, acc] : [acc],
      ];
    },
  };
}
