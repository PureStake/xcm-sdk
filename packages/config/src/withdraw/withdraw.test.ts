import { createWithdrawBuilder } from './withdraw';

describe('withdraw', () => {
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const withdraw = createWithdrawBuilder();

  describe('xTokens', () => {
    const cfg = withdraw.xTokens({
      balance: '<BALANCE>' as any,
      destination: { parachainId: 1000 } as any,
      existentialDeposit: 100_000_000_000,
      feePerWeight: 8,
    });

    it('should be correct withdraw config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get params with parachain id', () => {
      expect(cfg.getParams(account)).toMatchSnapshot();
    });

    it('should get params without parachain id', () => {
      const cfg2 = withdraw.xTokens({
        balance: '<BALANCE>' as any,
        destination: {} as any,
        existentialDeposit: 100_000_000_000,
        feePerWeight: 8,
      });

      expect(cfg2.getParams(account)).toMatchSnapshot();
    });
  });
});
