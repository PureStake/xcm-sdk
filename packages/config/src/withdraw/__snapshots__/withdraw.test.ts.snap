// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withdraw xTokens should be correct withdraw config 1`] = `
Object {
  "balance": "<BALANCE>",
  "destination": Object {
    "parachainId": 1000,
  },
  "feePerWeight": 8,
  "getParams": [Function],
  "sourceMinBalance": undefined,
  "weight": 4000000000,
}
`;

exports[`withdraw xTokens should get params with parachain id 1`] = `
Array [
  1,
  Array [
    "0x00000003e8",
    "0x01ef46c7649270c912704fb09b75097f6e32208b8500",
  ],
]
`;

exports[`withdraw xTokens should get params without parachain id 1`] = `
Array [
  1,
  Array [
    "0x01ef46c7649270c912704fb09b75097f6e32208b8500",
  ],
]
`;
